app/electron/main.ts

```diff
@@ -383,9 +393,81 @@ function startElecron() {
     });
 
     if (!useExternalServer) {
+      const runningHeadlamp = await getRunningHeadlampPIDs();
+      let shouldWaitForKill = true;
+
+      if (!!runningHeadlamp) {
+        const resp = dialog.showMessageBoxSync(mainWindow, {
+          // Avoiding mentioning Headlamp here because it may run under a different name depending on branding (plugins).
+          title: i18n.t('Another process is running'),
+          message: i18n.t(
+            'Looks like another process is already running. Continue by stopping that process automatically, or quit?'
+          ),
+          type: 'question',
+          buttons: [i18n.t('Continue'), i18n.t('Quit')],
+        });
+
+        if (resp === 0) {
+          runningHeadlamp.forEach(pid => {
+            try {
+              killProcess(pid);
+            } catch (e) {
+              console.log(`Failed to quit headlamp-servere:`, e.message);
+              shouldWaitForKill = false;
+            }
+          });
+        } else {
+          mainWindow.close();
+          return;
+        }
+      }
+
+      // If we reach here, then we attempted to kill headlamp-server. Let's make sure it's killed
+      // before starting our own, or else we may end up in a race condition (failing to start the
+      // new one before the existing one is fully killed).
+      if (!!runningHeadlamp && shouldWaitForKill) {
+        let stillRunning = true;
+        let timeWaited = 0;
+        const maxWaitTime = 3000; // ms
+        // @todo: Use an iterative back-off strategy for the wait (so we can start by waiting for shorter times).
+        for (let tries = 1; timeWaited < maxWaitTime && stillRunning; tries++) {
+          console.debug(
+            `Checking if Headlamp is still running after we asked it to be killed; ${tries} ${timeWaited}/${maxWaitTime}ms wait.`
+          );
+
+          // Wait (10 * powers of 2) ms with a max of 250 ms
+          const waitTime = Math.min(10 * tries ** 2, 250); // ms
```

Comment:
Cool... So now it's: 10, 40, 90, 160, 250. That seems good to me.

Reply:
More like 10, 40, 80, 160, 250.